src/land/smap9km_ssm2ioda.py

```diff
@@ -0,0 +1,173 @@
+            self.varDict[iodavar]['valKey'] = iodavar, iconv.OvalName()
+            self.varDict[iodavar]['errKey'] = iodavar, iconv.OerrName()
+            self.varDict[iodavar]['qcKey'] = iodavar, iconv.OqcName()
+            self.varAttrs[iodavar, iconv.OvalName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OerrName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OqcName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OvalName()]['units'] = 'm3 m-3'
+            self.varAttrs[iodavar, iconv.OerrName()]['units'] = 'm3 m-3'
+            self.varAttrs[iodavar, iconv.OqcName()]['units'] = 'm3 m-3'
+
+        # open input file name
+        ncd = nc.Dataset(self.filename, 'r')
+        # set and get global attributes
+        self.satellite = "SMAP"
+        self.sensor = "radar and radiometer"
+        AttrData["platform"] = self.satellite
+        AttrData["sensor"] = self.sensor
+
+        data = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['soil_moisture'][:]
+        vals = data[:].ravel()
+        _FillValue = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['soil_moisture'].getncattr('_FillValue')
+        valid_max = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['soil_moisture'].getncattr('valid_max')
+        valid_min = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['soil_moisture'].getncattr('valid_min')
+
+        lats = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['latitude'][:].ravel()
+        lons = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['longitude'][:].ravel()
+        errs = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['soil_moisture_error'][:].ravel()
+        qflg = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['retrieval_qual_flag'][:].ravel()
+        sflg = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['surface_flag'][:].ravel()
+        vegop = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['vegetation_opacity'][:].ravel()
+
+        times = np.empty_like(vals, dtype=object)
+
+        if self.mask == "maskout":
+            with np.errstate(invalid='ignore'):
+                mask = (vals > valid_min) & (vals < valid_max)
+            vals = vals[mask]
+            lats = lats[mask]
+            lons = lons[mask]
+            errs = errs[mask]
+            qflg = qflg[mask]
+            sflg = sflg[mask]
+            vegop = vegop[mask]
+            times = times[mask]
+
+        # get datetime from filename
+        str_split = self.filename.split("_")
+        str_datetime = str_split[7]
+        my_datetime = datetime.strptime(str_datetime, "%Y%m%dT%H%M%S")
+        base_datetime = my_datetime.strftime('%Y-%m-%dT%H:%M:%SZ')
+        vals = vals.astype('float32')
+        lats = lats.astype('float32')
+        lons = lons.astype('float32')
+        errs = errs.astype('float32')
+        qflg = qflg.astype('int32')
```

Comment:
qflag values are 0, 1,2, .... 13, and  sflg values are 0, 32, 128, 131, 135, 160, 163, 167, ...... if I used ncdump to take a look. If I used h5dump to look at SMAP_L2_SM_P_E_36365_D_20211122T013945_R18240_001.h5, these flag data are H5T_STD_U16LE - 16-bit little-endian bitfields. Now I am not sure if python code can read them directly and convert them to integer. @srherbener @rhoneyager, do you have any suggestions or experiences? I directly use  qflg = ncd.groups['Soil_Moisture_Retrieval_Data'].variables['retrieval_qual_flag'][:].ravel() with ncd = nc.Dataset(self.filename, 'r') and  import netCDF4 as nc (netCDF4 library).

Reply:
I am not sure if I need a special procedure to process H5T_STD_U16LE in python, any suggestions and/or expereince can share with me @CoryMartin-NOAA?

Reply:
Simply converting them to integer will be enough for us to get started on the DA. However, this means that we will need to follow the SMAP team's recommendations on which data to reject. There's quite a bit of evidence that they reject too much data, so eventually we may want to determine our own rules for what to reject - and this will require examining the individual bits. 

If the UFO can process bit-level QC flags, then I recommend we do the work now to include this information in the IODA files. If the UFO can't process bit level flags, let's stick with integer, and create an issue noting this will need to be resolved eventually (I'm assuming there are many other obs using bit-based QC).

Reply:
FYI: the netCDF4 python package should read the H5T_STD_U16LE data in the file as integers. That's the beauty of the HDF5 data model is that you don't need to worry about these kinds of details about the file.